src/Lookup/DispatchingEntityLookup.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Lookup;
+
+use Wikibase\DataModel\Entity\EntityDocument;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikimedia\Assert\Assert;
+use Wikimedia\Assert\ParameterAssertionException;
+
+/**
+ * Delegates lookup to the repository-specific EntityLookup
+ * based on the name of the repository EntityId is coming from.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class DispatchingEntityLookup implements EntityLookup {
+
+	/**
+	 * @var EntityLookup[]
+	 */
+	private $lookups;
+
+	/**
+	 * @param EntityLookup[] $lookups associative array with repository names (strings) as keys
+	 *                                and EntityLookup objects as values. Empty-string key
+	 *                                defines lookup for the local repository.
+	 *
+	 * @throws ParameterAssertionException
+	 */
+	public function __construct( array $lookups ) {
+		Assert::parameter(
+			!empty( $lookups ) && array_key_exists( '', $lookups ),
+			'$lookups',
+			'must must not be empty and must contain an empty-string key'
+		);
+		Assert::parameterElementType( EntityLookup::class, $lookups, '$lookups' );
+		Assert::parameterElementType( 'string', array_keys( $lookups ), 'array_keys( $lookups )' );
+		foreach ( array_keys( $lookups ) as $repositoryName ) {
+			Assert::parameter(
+				strpos( $repositoryName, ':' ) === false,
+				'array_keys( $lookups )',
+				'must not contain strings including colons'
+			);
+		}
+		$this->lookups = $lookups;
+	}
+
+	/**
+	 * @see EntityLookup::getEntity
+	 *
+	 * @param EntityId $entityId
+	 *
+	 * @return null|EntityDocument
+	 * @throws EntityLookupException
+	 * @throws UnknownForeignRepositoryException
+	 */
+	public function getEntity( EntityId $entityId ) {
+		$lookup = $this->getLookupForEntityId( $entityId );
+		return $lookup->getEntity( $entityId );
+	}
+
+	/**
+	 * @see EntityLookup::hasEntity
+	 *
+	 * @param EntityId $entityId
+	 *
+	 * @return bool
+	 * @throws EntityLookupException
+	 * @throws UnknownForeignRepositoryException
+	 */
+	public function hasEntity( EntityId $entityId ) {
+		$lookup = $this->getLookupForEntityId( $entityId );
+		return $lookup->hasEntity( $entityId );
+	}
+
+	/**
+	 * @param EntityId $entityId
+	 *
+	 * @return EntityLookup
+	 * @throws UnknownForeignRepositoryException
+	 */
+	private function getLookupForEntityId( EntityId $entityId ) {
+		$repo = $entityId->getRepositoryName();
+		if ( !isset( $this->lookups[$repo] ) ) {
+			throw new UnknownForeignRepositoryException( $entityId );
```

Comment:
Just like in the other PR this should probably also be `$entityId->getRepositoryName()` instead